File: backend/src/main/java/com/wanderlust/service/PostService.java
package com.wanderlust.service;

import com.wanderlust.model.Post;
import com.wanderlust.repository.PostRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class PostService {

    private final PostRepository postRepo;

    public PostService(PostRepository postRepo) {
        this.postRepo = postRepo;
    }

    /* =====================
       READ OPERATIONS
       ===================== */

    // Existing behavior preserved (first page)
    public List<Post> getAllPosts() {
        return postRepo
            .findAll(PageRequest.of(0, 5))
            .getContent();
    }

    public List<Post> getFeaturedPosts() {
        return postRepo
            .findByFeaturedPostTrue(PageRequest.of(0, 5))
            .getContent();
    }

    public List<Post> getPostsByAuthor(String authorName) {
        return postRepo.findByAuthorNameIgnoreCase(authorName);
    }

    public List<Post> getPostsByCategory(String category) {
        return postRepo
            .findByCategoriesContainingIgnoreCase(
                category,
                PageRequest.of(0, 5)
            )
            .getContent();
    }

    public Post getPostById(String id) {
        return postRepo.findById(id).orElse(null);
    }

    /* =====================
       WRITE OPERATIONS
       ===================== */

    public Post createPost(Post post) {
        return postRepo.save(post);
    }

    public void deletePost(String id) {
        postRepo.deleteById(id);
    }
}

File: frontend/src/services/post-service.ts
// src/services/posts.service.ts

import axiosInstance from "@/helpers/axios-instance";

export const getAllPosts = async () => {
  const response = await axiosInstance.get("/posts");
  return response.data;
};

export const getFeaturedPosts = async () => {
  const response = await axiosInstance.get("/posts/featured");
  return response.data;
};

export const getPostsByCategory = async (category: string) => {
  const response = await axiosInstance.get(
    `/posts/category/${encodeURIComponent(category)}`
  );
  return response.data;
};

export const getPostsByAuthor = async (authorName: string) => {
  const response = await axiosInstance.get(
    `/posts/author/${encodeURIComponent(authorName)}`
  );
  return response.data;
};

export const getPostById = async (id: string) => {
  const response = await axiosInstance.get(`/posts/${id}`);
  return response.data;
};
